File: message_delete_tracker.py
from dataclasses import dataclass


@dataclass
class PendingDeleteMessage:
    content: str
    channel_name: str
    delete_after_seconds: int
    queued_at: float


class PendingDeleteTracker:
    def __init__(self, stale_seconds=60.0):
        self.stale_seconds = float(stale_seconds)
        self._entries = []

    def add(self, content, channel_name, delete_after_seconds, now):
        self._entries.append(
            PendingDeleteMessage(
                content=content,
                channel_name=channel_name,
                delete_after_seconds=int(delete_after_seconds),
                queued_at=float(now),
            )
        )

    def pop_matched(self, content, channel_name, now):
        self._prune(now)
        for idx, entry in enumerate(self._entries):
            if entry.content == content and entry.channel_name == channel_name:
                return self._entries.pop(idx)
        return None

    def pop_first_for_channel(self, channel_name, now):
        self._prune(now)
        for idx, entry in enumerate(self._entries):
            if entry.channel_name == channel_name:
                return self._entries.pop(idx)
        return None

    def _prune(self, now):
        now = float(now)
        self._entries = [
            entry
            for entry in self._entries
            if (now - entry.queued_at) <= self.stale_seconds
        ]

File: src/chat/message-delete-tracker.ts
/**
 * 削除予約メッセージの追跡
 */
export interface PendingDeleteMessage {
  content: string;
  channelName: string;
  deleteAfterSeconds: number;
  queuedAt: number;
}

export class PendingDeleteTracker {
  private readonly staleSeconds: number;
  private readonly queue: PendingDeleteMessage[] = [];

  constructor(staleSeconds = 90) {
    this.staleSeconds = staleSeconds;
  }

  add(
    content: string,
    channelName: string,
    deleteAfterSeconds: number,
    now: number
  ): void {
    this.queue.push({ content, channelName, deleteAfterSeconds, queuedAt: now });
  }

  popMatched(
    content: string,
    channelName: string,
    now: number
  ): PendingDeleteMessage | null {
    this._prune(now);
    const idx = this.queue.findIndex(
      (m) => m.content === content && m.channelName === channelName
    );
    if (idx === -1) return null;
    return this.queue.splice(idx, 1)[0];
  }

  popFirstForChannel(
    channelName: string,
    now: number
  ): PendingDeleteMessage | null {
    this._prune(now);
    const idx = this.queue.findIndex((m) => m.channelName === channelName);
    if (idx === -1) return null;
    return this.queue.splice(idx, 1)[0];
  }

  private _prune(now: number): void {
    for (let i = this.queue.length - 1; i >= 0; i--) {
      if (now - this.queue[i].queuedAt > this.staleSeconds) {
        this.queue.splice(i, 1);
      }
    }
  }
}
